Nonlinear carb absorption model

dm61 opened the pull request:
Includes a piecewise-linear carb absorption model, and updates to the dynamic carb absorption algorithm to enable operation with nonlinear carb absorption models in general

dm61 committed:
include piecewise-linear carb absorption model, and enable nonlinear carb absorption models

naromero77 commented:
Just for clarification. The default Loop model would go from the current DCA where the absorption time is 1.5X the user specified absorption to a piecewise linear carb absorption model with an absorption time equal to (1.0X)the user specified absorption time. The piecewise linear absorption model has quicker absorption in the beginning and then tails off towards the end.

Perhaps I am reading into the title of the PR too much. Where does the nonlinear part come in? Are the higher order terms in the absorption model? (e.g. quadratic or exponential)

dm61 commented:
> Just for clarification. The default Loop model would go from the current DCA where the absorption time is 1.5X the user specified absorption to a piecewise linear carb absorption model with an absorption time equal to (1.0X)the user specified absorption time. The piecewise linear absorption model has quicker absorption in the beginning and then tails off towards the end.

Correct. Importantly, however, dynamic carb absorption algorithm still works, which means that the absorption time may be stretched up to 1.5x the user specified value based on data observed.

> Perhaps I am reading into the title of the PR too much. Where does the nonlinear part come in? Are the higher order terms in the absorption model? (e.g. quadratic or exponential)

The amount of absorbed carbs is an integral of the absorption rate. Standard Loop model has a flat absorption rate model, and therefore the amount of absorbed cabs is a linear function of time. This PR includes a model where the absorption rate is a piecewise linear function of time, so that the amount of absorbed carbs as a function of time includes quadratic terms. In the PR model, both the absorption rate and the amount absorbed are nonlinear functions of time, hence 'nonlinear' in the title.

ps2 commented on LoopKit/CarbKit/CarbMath.swift:
```diff
@@ -425,6 +499,8 @@ fileprivate class CarbStatusBuilder<T: CarbEntry> {
     let carbohydrateSensitivityFactor: Double
 
+    /// The absorption time for this entry before any absorption is observed
+    let initialAbsorptionTime: TimeInterval
```
`entry.absorptionTime` already contains this.

dm61 replied:
`entry.absorptionTime` is optional, and `defaultAbsorptionTime` is not in `CarbStatusBuilder`, so I'd keep `initialAbsorptionTime` as is, no?

ps2 commented on LoopKit/CarbKit/CarbMath.swift:
```diff
@@ -490,12 +566,25 @@ fileprivate class CarbStatusBuilder<T: CarbEntry> {
     }
 
-    /// The maximum amount of time needed for the remaining entry grams to absorb, at the fixed minimum absorption rate
+    /// The amount of time needed for the remaining entry grams to absorb, at the currently estimated absorption rate
```
For the non-linear model, the future amount of time is not necessarily a consistent rate, right? So this comment should not reference a single fixed rate.

dm61 replied:
Agreed, will fix the comment

ps2 reviewed (commented):
This needs to be implemented as an optional model (no UI for selection needed; could be a feature flag).  Also, unit tests should be included for algorithm changes.

dm61 commented on LoopKit/CarbKit/CarbMath.swift:
```diff
@@ -9,7 +9,7 @@
 import HealthKit
 
-
+// dm61 TODO(?): update protocol to include percentRateAtPercentTime, refactor all carb absorption models to conform to the updated protocol, move protocol and carb absorption models into separate files, unify references to absorption model, make it easy to select the active carb absorption model (could even make the model user selectable, similar to what we do with insulin models?)
```
The code-level model selection could be as simple as `typealias AbsorptionModel = LinearAbsorption` as default, with alternatives such as `typealias AbsorptionModel = PiecewiseLinearAbsorption` commented out. Would this be acceptable?

ps2 replied:
Look at https://github.com/LoopKit/Loop/blob/dev/Common/FeatureFlags.swift, to see how this could be done as a feature flag.

ps2 replied:
Since this is a LoopKit level patch, providing a new absorption model argument to the CarbStore would probably be the proper approach, then we can pass the correct one in based on the Loop feature flag.

dm61 replied:
Awesome, thanks for the guidance and the example - exactly what I was looking for!

ps2 commented:
Waiting on an api level switch to switch between linear and nonlinear models. And also waiting for a version that does not modify future *rate* based on observed absorption.

dm61 commented:
> Waiting on an api level switch to switch between linear and nonlinear models. And also waiting for a version that does not modify future _rate_ based on observed absorption.

dm61 commented:
Uh, clicked on a wrong button, reopening

dm61 commented:
@ps2 I've refactored the new carb model to make it easy to choose the model and some model settings. The choice between the standard model and the new nonlinear model is a simple switch. A user can further code-customize the settings in CarbMath to enable the rate adaptation (which is disabled by default). I'd need to do some more testing before merging this to the nonlinear-carb-model branch, but would be great if you could take a quick look at the commit, and let me know if this is what you are expecting to see - thanks.  https://github.com/dm61/LoopKit/commit/12fb65c91ed5f5d26e72597221efed10ffbedb6e

ps2 commented:
@dm61 CarbKit should expose a new public enum, like

ps2 commented:
CarbKit should expose an api to let callers select which model to use. Something like:
```
public enum CarbAbsorptionModel {
    case linear
    case piecewiseLinear
}
```
And then modify `CarbStore`'s init to allow an instance of that enum to be passed in to configure it.

dm61 commented:
ok, thanks, let me try. Another question: since we want to keep the non-adaptive minimum absorption rate as default, I think we should also keep the initialAbsorptionTime as 1.5*(entered time) as default for both nonlinear and linear models. Otherwise, we would have a weird jump in predicted carb effect as soon as absorption starts. In other words, the default nonlinear model would have just a different shape but would otherwise be identical to the current linear model. Agreed?

ps2 commented:
Yes, I think that makes sense. Also, even if minimal absorption is observed, carbs should still expire within initialAbsorptionTime * 1.5.

dm61 commented:
Does the interface look ok now?  https://github.com/dm61/LoopKit/commit/150b1bc8b0cff260197d8584d5be9b9eb858d18a 

I've included and have tested three model options: .linear, .nonlinear and .adaptiveRateNonlinear. The initial absorption time is 1.5*(entered) for .linear and .nonlinear, 1.0*(entered) for the .adaptiveRateNonlinear. The .linear model is the same as the standard Loop model. The .nonlinear model is the same, except the rate is piecewise linear. The adaptive rate model, after a standby interval, modifies future rate according to observed relative rate. The standby interval has been extended to a fraction of the entered absorption time to mitigate potentially noise-sensitive, too fast changes in dynamic absorption time during early absorption (as observed by some people in testing).

dm61 committed:
refactor carb model and add model selection interface

ps2 commented:
Tests are currently failing. Also it would be nice to have some tests covering the new nonlinear model.

dm61 committed:
unit tests for linear model

dm61 committed:
unit tests for nonlinear models

dm61 committed:
test errors in InsulinMathTests

dm61 committed:
Revert "test errors in InsulinMathTests"

This reverts commit 11dae6c0902ade3495aed0b24d801ba5586595b0.

dm61 commented:
I've updated all existing carb math tests, and have added similar unit tests for dynamic absorption with the piecewise linear model, as well as a partial-absorption test with the adaptive-rate option, which is really the only case when the adaptive-rate and standard-rate models differ.

_Unrelated note_: when I ran all unit LoopKit tests on my Xcode. couple of tests failed in  InsulinMathTest. I then tried to "fix" those just to get all tests to pass, with an intention of alerting you about this. But, Travis failed the "fixed" version, so I reverted those changes. Not really sure what's going on there - maybe the simulator I used or the pump assumed were different for me compared to settings in Travis (?)

PieterGit commented:
@dm61 or @ps2  I think I would like see if this carb absorption improves the blood sugars. Can you elaborate a bit more when this new setting is useful (and when not). I what case I should consider ro switch to the new carb absportion model.

Kdisimone commented:
@PieterGit Or go to Loopdocs now and find the information and links ;) https://loopkit.github.io/loopdocs/faqs/branch-faqs/#carb-model-changes-in-dev

Kdisimone commented:
(I don't like rewriting Loopdocs until features are merged into master whenever possible. I currently have to do some dev explanations just because of pod looping...but I want them to be in places clearly marked that they are dev related...so updating the main loopdocs carb absorption description is premature right now. I think once things are merged into master, then some of the docs updates will be more permanent)
